### ISSUE TITLE
Support multiple signature types

### ISSUE THREAD

**Issue by fgs-dsykes** [ralphje/signify#16]
As far as I can tell, this currently only looks at the signature at index 0, which includes a certificate chain and optionally a timestamp certificate chain. It is within spec for files to have an additional signature. An example is McAfee products which are signed with both a SHA1 and a SHA256 signature. Windows SignTool (https://docs.microsoft.com/en-us/windows/win32/seccrypto/signtool) does show and verify these signatures when ran with the /all parameter. As far as I can tell, Signify only handles the default signature (index: 0).

McAfee's definition updates (https://www.mcafee.com/enterprise/en-us/downloads/security-updates.html), which are distributed as executables with a SHA1 signature at index 0 and a SHA256 signature at index 1. As far as I can tell, signify is only parsing and verifying the SHA1, which is the least secure method available. It would be preferable to either parse the SHA256, or ideally, both signatures.

Am I simply missing some functionality in this? Is there any provision for this using the existing code? If not, what would be required to add it.

**Comment by ralphje** [comment]
We appear to do this in https://github.com/ralphje/signify/blob/master/signify/signed_pe.py#L256-L267 as at least one of the SignedDatas must be valid. I'm not sure that this is the correct approach though, as you are correct that if there are both a SHA-1 and SHA-256 certificate, we may want to prefer only checking the SHA-256 one.

Do you know how we should approach this, and how Windows handles when either one of the signatures is invalid?

**Comment by minisephirot** [comment]
I've been actively trying to parse binaries's multiples signatures with signify without much success.
The code you mentioned in signed_pe.py seems correct to me, but signed_datas yield only 1 object thus ignoring other signers.
 
Window considers SHA1 signatures to be deprecated unless it has been countersigned. 
SHA256 should be prefered but the various tools available on W10 only checks the first signing (just like your lib).

Can the lib pull multiples signing without heavy modification of source ? 
PS : I added a dump of various tools that checks signatures on windows (signcheck, get-authenticodesignature, my POC using signify)
[dump.txt](https://github.com/ralphje/signify/files/6028195/dump.txt)

**Comment by ralphje** [comment]
The issue seems to originate from a totally different use case: one SignerInfo with multiple certificates. This seems to be handled by adding an unauthenticated attribute in OID 1.3.6.1.4.1.311.2.4.1 to the SignerInfo class. We aren't currently parsing this, though we probably should.

**Comment by ralphje** [comment]
Could you verify that the above commits yields more expected results?

**Comment by minisephirot** [comment]
Hey @ralphje , you were on point with mutliples signatures beeing nesteds. It is working far better than the sysinternal tool now :
[MicrosoftEdgeSetupBeta (2).txt](https://github.com/ralphje/signify/files/6067299/MicrosoftEdgeSetupBeta.2.txt)
Thanks !

**Comment by ralphje** [comment]
This issue was still open, since I wanted to research how Microsoft handles the case of multiple signatures where only one is correct. Basically, the Microsoft default is only checking the first signature (at least in `sigcheck.exe`, `Get-AuthenticodeSignature` and `signtool` without `/all`). The other option is `signtool` with `/all`, which will show you that one is correct and the other isn't -- it doesn't provide any further guidance on what to do with that. Note that the `/as` option of `signtool` speaks about a 'primary' signature, whichever that may mean, but I gather this is what they mean with that.

Since there is no clear concensus about this, I have added a `multi_verify_mode` argument to allow customization by the caller, allowing the 'first' signature, 'all' signatures or 'any' signature to verify, and in spirit of your remarks, I have added the option to verify the 'best' signature as well.

For now, the Signify default remains to be 'any'.
